sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,864 @@
+        if ack_id not in self._ack_map:
+            self._ack_map[ack_id] = SendMessageErrorOptions(
+                error_detail=AckMessageError(name="", message="timeout to receive ack message")
+            )
+        try:
+            self._send_message(message, **kwargs)
+        except Exception as e:
+            self._ack_map.pop(ack_id)
+            raise e
+
+        with self._ack_map[ack_id].cv:
+            self._ack_map[ack_id].cv.wait(self._ack_timeout)
+            options = self._ack_map.pop(ack_id)
+            if options.error_detail is not None:
+                raise SendMessageError(
+                    message="Failed to send message.", ack_id=options.ack_id, error_detail=options.error_detail
+                )
+
+    def _get_or_add_group(self, name: str) -> WebPubSubGroup:
+        if name not in self._group_map:
+            self._group_map[name] = WebPubSubGroup(name=name)
+        return self._group_map[name]
+
+    @distributed_trace
+    def join_group(self, group_name: str, **kwargs: Any) -> None:
+        """Join the client to group.
+
+        :param group_name: The group name. Required.
+        :type group_name: str.
+        :keyword int ack_id: The optional ackId. If not specified, client will generate one.
+        """
+
+        def join_group_attempt():
+            group = self._get_or_add_group(group_name)
+            self._join_group_core(group_name, **kwargs)
+            group.is_joined = True
+
+        self._retry(join_group_attempt)
+
+    def _join_group_core(self, group_name: str, **kwargs: Any) -> None:
+        ack_id = kwargs.pop("ack_id", None)
+        self._send_message_with_ack_id(
+            message_provider=lambda id: JoinGroupMessage(group=group_name, ack_id=id),
+            ack_id=ack_id,
+            **kwargs,
+        )
+
+    @distributed_trace
+    def leave_group(self, group_name: str, **kwargs: Any) -> None:
+        """Leave the client from group
+        :param group_name: The group name. Required.
+        :type group_name: str.
+        :keyword int ack_id: The optional ackId. If not specified, client will generate one.
+        """
+
+        def leave_group_attempt():
+            group = self._get_or_add_group(group_name)
+            ack_id = kwargs.pop("ack_id", None)
+            self._send_message_with_ack_id(
+                message_provider=lambda id: LeaveGroupMessage(group=group_name, ack_id=id),
+                ack_id=ack_id,
+                **kwargs,
+            )
+            group.is_joined = False
+
+        self._retry(leave_group_attempt)
+
+    @distributed_trace
+    def send_event(
+        self,
+        event_name: str,
+        content: Any,
+        data_type: Union[WebPubSubDataType, str],
+        **kwargs: Any,
+    ) -> None:
+        """Send custom event to server
+        :param event_name: The event name. Required.
+        :type event_name: str.
+        :param content: The data content. Required.
+        :type content: Any.
```

Comment:
Are we able to give more guidance (either in the description or the type hint) as to what I should pass in here?

Reply:
updated